File: readleakage.jl
function read_leakage(Δm, Δl, ℓ, ℓ′)


  rleaks1 = OffsetArray(zeros(Float64,2*Δm+1,2*Δl+1,2*ℓ+1),-Δm:Δm,-Δl:Δl,-ℓ:ℓ)
  rleaks2 = OffsetArray(zeros(Float64,2*Δm+1,2*Δl+1,2*ℓ′+1),-Δm:Δm,-Δl:Δl,-ℓ′:ℓ′)
  horleaks1 = OffsetArray(zeros(Float64,2*Δm+1,2*Δl+1,2*ℓ+1),-Δm:Δm,-Δl:Δl,-ℓ:ℓ)
  horleaks2 = OffsetArray(zeros(Float64,2*Δm+1,2*Δl+1,2*ℓ′+1),-Δm:Δm,-Δl:Δl,-ℓ′:ℓ′)

  dl_mat=6
  dm_mat=15

  for ii in 1:2
    if(ii==1) lp=ℓ;end;
    if(ii==2) lp=ℓ′;end;
    ind0 = Int(lp*(lp+1)/2)

    for mp in 0:lp
     ind = ind0 + mp + 1
     for j = -Δl:Δl
      l = lp + j

      for i in -Δm:Δm

       m = mp + i
       if (abs(m) > l) continue;end;

       L_ii = 0.5*sign(m)*ci_mat[(abs(m)-abs(mp))+dm_mat+1,l-lp+dl_mat+1,ind]
       L_rr = 0.5*cr_mat[abs(m)-abs(mp)+dm_mat+1,l-lp+dl_mat+1,ind]

       if (ii==1) 
	       rleaks1[i,j,mp] = L_rr + L_ii
	       rleaks1[i,j,-mp] = L_rr + sign(mp)* L_ii
       else
	       rleaks2[i,j,mp] = L_rr + L_ii
	       rleaks2[i,j,-mp] = L_rr + sign(mp)* L_ii
       end

       L_ii = 0.5*sign(m)*hi_mat[abs(m)-abs(mp)+dm_mat+1,l-lp+dl_mat+1,ind]
       L_rr = 0.5*hr_mat[abs(m)-abs(mp)+dm_mat+1,l-lp+dl_mat+1,ind]

       if (ii==1) 
	       horleaks1[i,j,mp] = L_rr + L_ii
	       horleaks1[i,j,-mp] = L_rr + sign(mp)* L_ii
       else
	       horleaks2[i,j,mp] = L_rr + L_ii
	       horleaks2[i,j,-mp] = L_rr + sign(mp)* L_ii
       end

      end
     end
    end
  end

  return rleaks1, horleaks1, rleaks2, horleaks2
end

#-------------------------------------------------------

File: readleakage.py
# {{{ Library Importsj
from heliosPy import iofuncs as cio
from astropy.io import fits
import numpy as np
import time
# }}} imports

# {{{ Global variables
dl_mat = 6
dm_mat = 15
lmax = 249
# }}} global vars


def sign(a, b):
    """Returns the the value of a with sign of b

    Inputs:
    -------
    a - float
        variable whose value is returned
    b - float
        variable whose sign is returned

    Returns:
    --------
    abs(a) if b >= 0
    -abs(a) if b < 0

    """
    if b < 0:
        return -abs(a)
    else:
        return abs(a)


if __name__ == "__main__":
    t1 = time.time()
    data_dir = "/scratch/seismogroup/code/leakvw0/default/"
    write_dir = "/scratch/g.samarth/HMIDATA/leakmat/"
    rmat1 = fits.open(data_dir + "vradsum/leakrlist.vradsum.fits")[0].data
    rmat2 = fits.open(data_dir + "vradsum/leakilist.vradsum.fits")[0].data
    hormat1 = fits.open(data_dir + "vhorsum/leakrlist.vhorsum.fits")[0].data
    hormat2 = fits.open(data_dir + "vhorsum/leakilist.vhorsum.fits")[0].data

    rleaks = np.zeros((2*dm_mat+1, 2*dl_mat+1,
                       lmax+1, 2*lmax + 1), dtype=complex)
    horleaks = np.zeros((2*dm_mat+1, 2*dl_mat+1,
                         lmax+1, 2*lmax + 1), dtype=complex)

    for lp in range(lmax+1):
        ind0 = int(lp*(lp+1)/2)
        for mp in range(lp+1):
            ind = ind0 + mp
            for j in range(2*dl_mat + 1):
                ell = lp + j - dl_mat
                for i in range(2*dm_mat + 1):
                    m = mp + i - dm_mat
                    if abs(m) <= ell:
                        L_ii = 0.5 * sign(1, m) *\
                            rmat2[ind, ell-lp+dl_mat, abs(m)-abs(mp)+dm_mat]
                        L_rr = 0.5 * rmat1[ind, ell-lp+dl_mat,
                                           abs(m)-abs(mp)+dm_mat]
                        rleaks[i, j, lp, mp+lmax] = L_rr + L_ii
                        rleaks[i, j, lp, -mp+lmax] = L_rr + sign(1, mp)*L_ii

                        L_ii = 0.5 * sign(1, m) *\
                            hormat2[ind, ell-lp+dl_mat, abs(m)-abs(mp)+dm_mat]
                        L_rr = 0.5 * hormat1[ind, ell-lp+dl_mat,
                                             abs(m)-abs(mp)+dm_mat]
                        horleaks[i, j, lp, mp+lmax] = L_rr + L_ii
                        horleaks[i, j, lp, -mp+lmax] = L_rr + sign(1, mp)*L_ii
    t2 = time.time()
    print(f" Time taken for computation = {(t2 - t1)/60:.3f} minutes")
    cio.writefitsfile(rleaks.real, write_dir + 'rleaks1.fits')
    cio.writefitsfile(rleaks.imag, write_dir + 'rleaks2.fits')
    cio.writefitsfile(horleaks.real, write_dir + 'horleaks1.fits')
    cio.writefitsfile(horleaks.imag, write_dir + 'horleaks2.fits')
    print("writing complete")
